src/smtpproto/protocol.py

```diff
@@ -291,9 +309,10 @@ def mail(self, sender: Union[str, Address]) -> None:
         args = []
         if '8BITMIME' in self._extensions:
             args.append('BODY=8BITMIME')
+        if 'SMTPUTF8' in self._extensions:
+            args.append('SMTPUTF8')
```

Comment:
There's one thing I just realized now may also be a problem (sorry, would have noted it earlier if I'd realized) — from the RFC:

>    Because there is no guarantee that a next-hop SMTP server will
>    support the SMTPUTF8 extension, use of the SMTPUTF8 extension always
>    carries a risk of transmission failure.  In fact, during the early
>    stages of deployment for the SMTPUTF8 extension, the risk will be
>    quite high.  Hence, there is a distinct near-term advantage for
>    ASCII-only messages to be sent without using this extension.  The
>    long-term advantage of casting ASCII [ASCII] characters (0x7f and
>    below) as UTF-8 form is that it permits pure-Unicode environments.

So I think in practice, many (most?) servers will support SMTPUTF8, but we shouldn't use it unless the message requires it.

The stdlib smtplib module handles this by testing all of the address to see if any of them are non-ascii: https://github.com/python/cpython/blob/master/Lib/smtplib.py#L953. That's at the whole message level though, so I'm not sure how we could make that determination in smtpproto.

Reply:
If one of the addresses has non-ASCII characters in it, wouldn't the message fail to reach its destination anyway in such an event? How does not using SMTPUTF8 help in that case?

Reply:
There is another issue: if the SMTPUTF8 extension is not present, the client should use IDNA encoding to transform any non-ASCII domain name to ASCII (if I'm interpreting [RFC 6531](https://tools.ietf.org/html/rfc6531#section-3.7) correctly).

Reply:
> If one of the addresses has non-ASCII characters in it, wouldn't the message fail to reach its destination anyway in such an event? How does not using SMTPUTF8 help in that case?

The issue is for addresses that are only ASCII, on servers that support SMTPUTF8. In that case I think as things are currently, smtpproto will use the SMTPUTF8 option, because the server supports it. But we should only use it if the message needs it.

> There is another issue: if the SMTPUTF8 extension is not present, the client should use IDNA encoding to transform any non-ASCII domain name to ASCII (if I'm interpreting RFC 6531 correctly).

That may be the case (need to spend more time looking at the RFC). If so though, I'm pretty sure it's not something that's handled in aiosmtplib or the smtplib stdlib module.